### ISSUE TITLE
Do I need to preprocess my image if I have a preprocessing layer in my network model?

### ISSUE THREAD

**Issue by TrueWodzu** [Dobiasd/frugally-deep#272]
Hi Tobias,

I am just starting, I've created a model using Keras. In my model I am adding preprocessing layer:
` x = layers.experimental.preprocessing.Rescaling(1.0 / 255)(x)`
Will my model will work out of the box with this layer, or do I need to rescale my image in C++ before passing it to my model?

Thanks :)

**Comment by Dobiasd** [comment]
Hi TrueWodzu,

Frugally-deep does not support the [`Rescaling` layer](https://www.tensorflow.org/api_docs/python/tf/keras/layers/experimental/preprocessing/Rescaling), or any other experimental layers. Once `Rescaling` is added to the normal Keras API, I might add support for it. Until then, in order to use your model with frugally-deep, you have to remove the `Rescaling` layer and do this part manually.

**Comment by TrueWodzu** [comment]
Hi Tobias,

Thank you for a very fast answer:) I am sure you have enormous experience with this. If that is not too much problem, could you show me how can I rescale my OpenCV image after loading from file and before passing to my network?

Should I use `cv::normalize()` to do this?

**Comment by Dobiasd** [comment]
Since the `Rescale` layer applies a fixed `scale` and `offset`, I suggest do use [`cv::Mat::convertTo`](https://docs.opencv.org/3.4/d3/d63/classcv_1_1Mat.html#adf88c60c5b4980e05bb556080916978b). `alpha` corresponds to `scale`, `beta` corresponds to `offset`.

**Comment by TrueWodzu** [comment]
Thank you.

**Comment by TrueWodzu** [comment]
Tobias,

Maybe I am stupid, but isn't scaling already included in this call? 

    const auto input = fdeep::tensor_from_bytes(image.ptr(),
        static_cast<std::size_t>(image.rows),
        static_cast<std::size_t>(image.cols),
        static_cast<std::size_t>(image.channels()),
        **0.0f, 1.0f**);

That would mean that I don't need to use `cv::Mat::convertTo` .

**Comment by Dobiasd** [comment]
Yes, you can use this too. You just need to derive the values for `low` and `high` for `tensor_from_bytes` from the values you're using for `scale` and `offset` for `Rescaling`.

**Comment by TrueWodzu** [comment]
Thanks,

I came to the point that everything works, however I am getting different prediction values in comparison to Keras. These differences are not large, the classification works exactly the same, however I've investigated this after reading your FAQ and just wanted to ask you about your experience with this.

1. Keras interpolation method ="nearest" when resizing an image works differently than `OpenCv cv::resize`. Have you noticed that in your practice?
2. I got different tensor shape .

In Keras I am getting: `shape=(1, 398, 720, 3)` due to this code:

    img = keras.preprocessing.image.load_img(test_image_file_name, target_size=image_size)
    img_array = keras.preprocessing.image.img_to_array(img)
    img_array = tf.expand_dims(img_array, 0)  # This adds "1" at the beginning of the tensor.
    predictions = model.predict(img_array)

In frugally I am getting: `(398, 720, 3)`

I guess, in Keras this "1" is added in case I would post a multiple images at once to model.predict() right?

**Comment by Dobiasd** [comment]
1. I don't know about the difference in the implementation of the "nearest" method in Keras and OpenCV, but yeah, they might be different. However, in general, for shrinking images, "nearest" is not the best option. "linear" is not a lot better, but `INTER_AREA` usually is good, because it avoids Moiré patterns, image-scaling attacks, etc. Or `LANCZOS`. So if you can afford it (performance-wise), they might be a good choice.

2. You're right. `expand_dims` adds an outer dimension (size `1` in that case), because the model can take multiple images for batch processing. [In frugally-deep you can do something similar by using fdeep::model::predict`.](https://github.com/Dobiasd/frugally-deep/blob/master/FAQ.md#does-frugally-deep-support-multiple-cpus)

**Comment by TrueWodzu** [comment]
Thank you, I will test different resizing methods and their impact on accuracy:)
